File: experiment-results/nodejs/instrumented/610.dynamic-html-otel/610.dynamic-html-otel_code/nodejs/12/docker/function.js
const Mustache = require('mustache');
const fs = require('fs');
const path = require('path');

const { BasicTracerProvider } = require('@opentelemetry/sdk-trace-base');
const { Resource } = require('@opentelemetry/resources');
const { SEMRESATTRS_SERVICE_NAME } = require('@opentelemetry/semantic-conventions');
const { OTLPTraceExporter } = require('@opentelemetry/exporter-trace-otlp-grpc');
const { SimpleSpanProcessor } = require('@opentelemetry/sdk-trace-base');
const opentelemetry = require('@opentelemetry/api');

const tracerProvider = new BasicTracerProvider({
  resource: new Resource({
    [SEMRESATTRS_SERVICE_NAME]: '610.dynamic-html-opentelemetry-nodejs',
  }),
});

const traceExporter = new OTLPTraceExporter({
  url: 'http://192.168.1.104:4317',
});

tracerProvider.addSpanProcessor(new SimpleSpanProcessor(traceExporter));
tracerProvider.register();

const tracer = tracerProvider.getTracer('nodejs-tracer');

function random(b, e, parentSpan) {
  const ctx = opentelemetry.trace.setSpan(
    opentelemetry.context.active(),
    parentSpan,
  );

  const span = tracer.startSpan('random', undefined, ctx);

  const result = Math.round(Math.random() * (e - b) + b);
  span.end();
  return result;
}

exports.handler = async function(event) {
  const span = tracer.startSpan('handler');

  var random_numbers = new Array(event.random_len);
  for (var i = 0; i < event.random_len; ++i) {
    random_numbers[i] = random(0, 100, span);
  }

  var input = {
    cur_time: new Date().toLocaleString(),
    username: event.username,
    random_numbers: random_numbers
  };
  span.setAttribute('username', event.username)
  span.setAttribute('random_numbers', random_numbers)

  var file = path.resolve(__dirname, 'templates', 'template.html');
  span.setAttribute('template_file', file);

  return new Promise((resolve, reject) => {
    fs.readFile(file, "utf-8", function(err, data) {
      if (err) {
        span.recordException(err);
        span.end();
        reject(err);
      } else {
        const rendered = Mustache.render(data, input);
        span.setAttribute('rendered.length', rendered.length);
        span.end();
        resolve(rendered);
      }
    });
  });
};

File: e2-task-based-experiment/benchmarks/600.opentelemetry/610.dynamic-html-otel/python/function.py
from datetime import datetime
from random import sample
from os import path
from time import time
import os
from opentelemetry import trace
from opentelemetry.exporter.otlp.proto.grpc.trace_exporter import OTLPSpanExporter
from opentelemetry.sdk.trace import TracerProvider
from opentelemetry.sdk.trace.export import SimpleSpanProcessor
from opentelemetry.sdk.resources import Resource

from jinja2 import Template

SCRIPT_DIR = path.abspath(path.join(path.dirname(__file__)))


resource = Resource(attributes={
    "service.name": "610.dynamic-html-opentelemetry"
})

trace.set_tracer_provider(TracerProvider(resource=resource))
tracer = trace.get_tracer("function")

# Setup the OTLP exporter to send traces to the collector using HTTP
otlp_exporter = OTLPSpanExporter(
    endpoint="http://192.168.1.109:4317",
    insecure=True
)

# Setup the span processor with the exporter
span_processor = SimpleSpanProcessor(otlp_exporter)
trace.get_tracer_provider().add_span_processor(span_processor)


def handler(event):
    span = tracer.start_span("handler")

    name = event.get('username')
    size = event.get('random_len')
    cur_time = datetime.now()

    span.set_attribute("username", name)
    span.set_attribute("random_len", size)

    random_numbers = sample(range(0, 1000000), size)
    template_path = os.path.join(SCRIPT_DIR, 'templates', 'template.html')

    with open(template_path, 'r') as file:
        template = Template(file.read())

    html = template.render(username=name, cur_time=cur_time, random_numbers=random_numbers)

    span.set_attribute("template_path", template_path)
    span.set_attribute("render_time", str(cur_time))

    span.end()
    return {'result': html}
